cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -246,6 +247,97 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
   return trajectory_nodes_list;
 }
 
+visualization_msgs::MarkerArray MapBuilderBridge::GetConstraintsList(
+    const ::cartographer::mapping::SparsePoseGraph::Constraint::Tag
+        wanted_constraint_tag) {
+  visualization_msgs::MarkerArray constraints_viz;
+
+  const auto trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+  std::vector<cartographer::transform::Rigid3d> submap_transforms;
+  for (auto& single_trajectory : trajectory_nodes) {
+    for (auto& node : single_trajectory) {
+      auto current_trajectory_transforms =
+          map_builder_.sparse_pose_graph()->GetSubmapTransforms(
+              node.constant_data->trajectory_id);
+      std::move(current_trajectory_transforms.begin(),
+                current_trajectory_transforms.end(),
+                std::back_inserter(submap_transforms));
+    }
+  }
+
+  int marker_id = 0;
+  ros::Time now = ros::Time::now();
+  for (const auto& constraint : constraints) {
+    visualization_msgs::Marker constraint_marker, residual_error_marker;
+
+    std_msgs::ColorRGBA color_constraint, color_error;
+
+    if (constraint.tag != wanted_constraint_tag) {
+      continue;
+    }
+
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTER_SUBMAP) {
+      // yellow
+      color_constraint.a = 1.0;
+      color_constraint.r = color_constraint.g = 1.0;
+      // cyan
+      color_error.a = 1.0;
+      color_error.b = color_error.g = 1.0;
+    } else {
+      color_constraint.a = color_error.a = 1.0;
+      color_constraint.g = 1.0;
+      color_error.b = 1.0;
+    }
+
+    constraint_marker = createVisualizationMarker(
+        marker_id++, visualization_msgs::Marker::LINE_STRIP, "constraint", now,
+        node_options_.map_frame);
+    constraint_marker.color = color_constraint;
+    residual_error_marker = createVisualizationMarker(
+        marker_id++, visualization_msgs::Marker::LINE_STRIP, "residual error",
+        now, node_options_.map_frame);
+    residual_error_marker.color = color_error;
+
+    geometry_msgs::Point submap_point, submap_pose_point, trajectory_node_point;
+
+    submap_point = ToGeometryMsgPoint(
+        submap_transforms[constraint.submap_id.submap_index].translation());
+    submap_pose_point = ToGeometryMsgPoint(
+        (submap_transforms[constraint.submap_id.submap_index] *
+         constraint.pose.zbar_ij)
+            .translation());
+    trajectory_node_point =
+        ToGeometryMsgPoint(trajectory_nodes[constraint.node_id.trajectory_id]
+                                           [constraint.node_id.node_index]
+                                               .pose.translation());
+
+    constraint_marker.points.push_back(submap_point);
+    constraint_marker.points.push_back(submap_pose_point);
+    residual_error_marker.points.push_back(submap_pose_point);
```

Comment:
You have a point there, @wohe. But maybe this explanation helps: Originally I wanted to see the two end points of a `constraint`. So the `submap` an the `trajectory_node`, because I was adding additional `constraints` and I wanted to see if it works.
Later I wanted to see where the `constraints` are "pulling" a `trajectory_node`. This is what the `residual_error_marker`. 
Since technically it is not only pulling a `trajectory_node` torwards a point on a `submap`, but also a `submap` torwards that `trajectory_node`, it might still be useful to see the `submap` a constraint is attached to (or: whether two `trajectory_nodes` are constrained to the same `submap`.) This is the `constraint_marker`. Now in most use cases the `residual_error_marker` might be the more practical, which is why I would keep them as seperate topics, but with Landmarks in the pipeline this might soon get more relevant again.